### PR TITLE
[LibGD] Add builder

### PR DESCRIPTION
**Pull request by clad26**
This is, as of yet, incomplete.
I just have this ``` build_tarballs.jl ``` for now and I want to check how it works on other systems.
How do I do that?

### CODE REVIEW

**Comment by clad26** [comment]
First problem is 
```
sources = [
    ArchiveSource("https://github.com/libgd/libgd/releases/download/gd-2.3.0/libgd-2.3.0.tar.gz")
]
```
There should be a uuid number here

**Comment by giordano** [comment]
https://docs.binarybuilder.org/stable/FAQ/#What-are-those-numbers-in-the-list-of-sources?-How-do-I-get-them?

**Comment by clad26** [comment]
@giordano Thanks!!! Works now

**Comment by clad26** [comment]
@giordano 
I think this might be ready to merge!
I'll use this as a dependency to build gnuplot

File: L/libgd/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,47 @@
+using BinaryBuilder, Pkg
+
+name = "libgd"
```
Comment (giordano) [inline L/libgd/build_tarballs.jl:3]:
```suggestion
name = "LibGD"
```
Note that you need to rename the directory to `LibGD`

File: L/libgd/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+# Collection of sources required to complete build
+sources = [
+    ArchiveSource("https://github.com/libgd/libgd/releases/download/gd-2.3.0/libgd-2.3.0.tar.gz", "32590e361a1ea6c93915d2448ab0041792c11bae7b18ee812514fe08b2c6a342")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/libgd-*/
+./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target}
+make
```
Comment (giordano) [inline L/libgd/build_tarballs.jl:15]:
```suggestion
make -j${nproc}
```

**Comment by clad26** [comment]
@giordano If possible, can you take a look at this commit?
https://github.com/JuliaPackaging/Yggdrasil/pull/2335/commits/06d11c87538c88aafbf8f1113fd66d6f67d02532
It's giving an error for two architectures

**Comment by giordano** [comment]
First of all, that means that `Libiconv_jll` must be another dependency, because libgd would link to it.  Secondly, try to add this option to `configure`: `--with-libiconv-prefix="${prefix}"`

File: L/libgd/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    LibraryProduct("libgd", :libgd),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="JpegTurbo_jll", uuid="aacddb02-875f-59d6-b918-886e6ef4fbf8")),
+    Dependency(PackageSpec(name="Zlib_jll", uuid="83775a58-1f1d-513f-b197-d71354ab007a")),
+    Dependency(PackageSpec(name="libpng_jll", uuid="b53b4c65-9356-5827-b1ea-8c7a1a84506f")),
+    Dependency(PackageSpec(name="Libtiff_jll", uuid="89763e89-9b03-5906-acba-b20f662cd828")),
+    Dependency(PackageSpec(name="Xorg_xorgproto_jll", uuid = "c4d99508-4286-5418-9131-c86396af500b")),
```
Comment (giordano) [inline L/libgd/build_tarballs.jl:44]:
```suggestion
    BuildDependency(PackageSpec(name="Xorg_xorgproto_jll", uuid = "c4d99508-4286-5418-9131-c86396af500b")),
```

File: L/libgd/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="JpegTurbo_jll", uuid="aacddb02-875f-59d6-b918-886e6ef4fbf8")),
+    Dependency(PackageSpec(name="Zlib_jll", uuid="83775a58-1f1d-513f-b197-d71354ab007a")),
+    Dependency(PackageSpec(name="libpng_jll", uuid="b53b4c65-9356-5827-b1ea-8c7a1a84506f")),
+    Dependency(PackageSpec(name="Libtiff_jll", uuid="89763e89-9b03-5906-acba-b20f662cd828")),
+    BuildDependency(PackageSpec(name="Xorg_xorgproto_jll", uuid = "c4d99508-4286-5418-9131-c86396af500b")),
+    Dependency(PackageSpec(name="Xorg_libXpm_jll", uuid = "1a3ddb2d-74e3-57f3-a27b-e9b16291b4f2")),
+    BuildDependency(PackageSpec(name="Libiconv_jll", uuid = "94ce4f54-9a6c-5748-9c1c-f9c7231a4531")),
```
Comment (giordano) [inline L/libgd/build_tarballs.jl:46]: libiconv would be needed at runtime, so must be a `Dependency`:
```suggestion
    Dependency(PackageSpec(name="Libiconv_jll", uuid = "94ce4f54-9a6c-5748-9c1c-f9c7231a4531")),
```

**Comment by clad26** [comment]
@giordano libiconv still gives an error

**Comment by clad26** [comment]
Is there some issue in libXpm and libwebp? or in libiconv itself?

File: L/libgd/build_tarballs.jl (hunk 1 above)
Comment (giordano) [reply L/libgd/build_tarballs.jl]: > Note that you need to rename the directory to `LibGD`

**Comment by clad26** [comment]
All of them erred... just by changing the name to LibGD....
Thats a bit confusing....😅

File: L/libgd/build_tarballs.jl (hunk 1 above)
Comment (clad26) [reply L/libgd/build_tarballs.jl]: Done!

Comment (clad26) [reply L/libgd/build_tarballs.jl]: Still error.... with libiconv

Comment (giordano) [reply L/libgd/build_tarballs.jl]: I'll have a look later

Comment (clad26) [reply L/libgd/build_tarballs.jl]: @giordano This is the only unbuilt dependency of gnuplot 5.4.1 I guess!
After this, I'll try that, and try linking it to [Gnuplot.jl](https://github.com/gcalderone/Gnuplot.jl) and [Gaston.jl](https://github.com/mbaz/Gaston.jl)
Seriously, thanks a lot!!! 😀

File: L/LibGD/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,51 @@
+using BinaryBuilder, Pkg
+
+name = "LibGD"
+version = v"2.3.0"
+
+# Collection of sources required to complete build
+sources = [
+    ArchiveSource("https://github.com/libgd/libgd/releases/download/gd-2.3.0/libgd-2.3.0.tar.gz", "32590e361a1ea6c93915d2448ab0041792c11bae7b18ee812514fe08b2c6a342")
```
Comment (fingolfin) [inline L/LibGD/build_tarballs.jl:8]:
```suggestion
    ArchiveSource("https://github.com/libgd/libgd/releases/download/gd-$(version)/libgd-$(version).tar.gz",
                  "32590e361a1ea6c93915d2448ab0041792c11bae7b18ee812514fe08b2c6a342")
```

**Comment by clad26** [comment]
@fingolfin Appreciate the suggestion!
I doubt that'll help with the current problem though!

**Comment by fingolfin** [comment]
The build error looks as if a `-liconv` linker flag is somehow missing. Indeed, here's one of the failed link commands in the x86_64 muslc build:
```
[18:23:19] /bin/sh ../libtool  --tag=CC   --mode=link cc -std=gnu99  -g -O2 -fvisibility=hidden -Wall   -o gd2togif gd2togif.o libgd.la  -L/workspace/destdir/lib -lz   -L/workspace/destdir/lib -lpng16 -lz   -L/workspace/destdir/lib -ljpeg   -L/workspace/destdir/lib -lXpm -lX11   -L/workspace/destdir/lib -ltiff    -lwebp
[18:23:19] libtool: link: cc -std=gnu99 -g -O2 -fvisibility=hidden -Wall -o .libs/gd2togif gd2togif.o  ./.libs/libgd.so -L/workspace/destdir/lib -lpng16 -lz -ljpeg /workspace/destdir/lib/libXpm.so /workspace/destdir/lib/libX11.so /workspace/destdir/lib/libtiff.so /workspace/destdir/lib/libwebp.so -Wl,-rpath -Wl,/workspace/destdir/lib
...
./.libs/libgd.so: undefined reference to `libiconv'
```
So no `-liconv` flag, while in the [x86_64 glibc build](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=8151&view=logs&j=38f3f99e-48ee-5298-7d91-5e943bf5d01b&t=5ae31a84-7b7f-52e6-2196-8f15fa5036ee), there is a `-liconv` flag.
```
[17:53:26] /bin/sh ../libtool  --tag=CC   --mode=link cc -std=gnu99  -g -O2 -fvisibility=hidden -Wall   -o gd2togif gd2togif.o libgd.la -liconv -lm   -L/workspace/destdir/lib -lz   -L/workspace/destdir/lib -lpng16 -lz   -L/workspace/destdir/lib -ljpeg   -L/workspace/destdir/lib -lXpm -lX11   -L/workspace/destdir/lib -ltiff    -lwebp
[17:53:27] libtool: link: cc -std=gnu99 -g -O2 -fvisibility=hidden -Wall -o .libs/gd2togif gd2togif.o /workspace/destdir/lib/libiconv.so -Wl,-rpath -Wl,/workspace/destdir/lib  ./.libs/libgd.so -L/workspace/destdir/lib -lpng16 -lz -ljpeg /workspace/destdir/lib/libXpm.so /workspace/destdir/lib/libX11.so /workspace/destdir/lib/libtiff.so /workspace/destdir/lib/libwebp.so -Wl,-rpath -Wl,/workspace/destdir/lib

```

Could the issue be caused by libiconv being somehow different on muslc systems, and/or a build system dependency being  picked up over that from the libiconv JLL? Because in the [i686 musl build](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=8151&view=logs&j=e8c8b252-e770-5e1c-dc43-368ae1d6cdff&t=42110e14-53ad-53b2-e425-388e4c88d382), everything links despite there not being a `-liconv`:
```
[17:53:27] /bin/sh ../libtool  --tag=CC   --mode=link cc -std=gnu99  -g -O2 -fvisibility=hidden -Wall   -o gd2togif gd2togif.o libgd.la /workspace/destdir/lib/libiconv.so -Wl,-rpath -Wl,/workspace/destdir/lib -L/workspace/destdir/lib -lz   -L/workspace/destdir/lib -lpng16 -lz   -L/workspace/destdir/lib -ljpeg   -L/workspace/destdir/lib -lXpm -lX11   -L/workspace/destdir/lib -ltiff    -lwebp
[17:53:27] libtool: link: cc -std=gnu99 -g -O2 -fvisibility=hidden -Wall -o .libs/gd2togif gd2togif.o /workspace/destdir/lib/libiconv.so -Wl,-rpath -Wl,/workspace/destdir/lib  ./.libs/libgd.so -L/workspace/destdir/lib -lpng16 -lz -ljpeg /workspace/destdir/lib/libXpm.so /workspace/destdir/lib/libX11.so /workspace/destdir/lib/libtiff.so /workspace/destdir/lib/libwebp.so -Wl,-rpath -Wl,/workspace/destdir/lib
```

**Comment by clad26** [comment]
@clad26 @fingolfin Any update to this?😅

**Comment by giordano** [comment]
I'm looking at this right now.  I'm very confused about what happens in `x86_64-linux-musl`.  I believe I found a cheap solution, but I'm trying to better figure out what's going wrong

**Pull request merged**